### ISSUE TITLE
Improving reconciliation against external identifiers

### ISSUE THREAD

**Issue by lozanaross** [OpenRefine/OpenRefine#6044]
This issue is a quick follow up from discussion about reconciliation with @wetneb prompted by a closer look here [#2083](https://github.com/OpenRefine/OpenRefine/issues/2083). Reconciling for external identifiers needs more attention as at the moment, it is not straightforward and requires some work-arounds. It could be part of the reworking of the whole reconciliation service dialog. 

Essentially it is important to provide option to "select any property, such as subclass of, or an external ID, against which to reconcile the cell value which is not necessarily a label."

@wetneb perhaps you can elaborate here on some further solutions / alternatives that can be considered.

**Comment by D063520** [comment]
I'm for this

**Comment by wetneb** [comment]
Here is an attempt to state the problem:

Given a dataset which contains some properties of entities to reconcile, but not their name, there is currently no simple way to reconcile those entities only based on those properties. This is a problem, because those properties might be able to identify the entities well enough on their own, without having to supply a name. For instance, if the properties values are unique in the reconciliation service (because they are designed as unique identifiers), then it's legitimate to expect that supplying those property values would suffice.

For instance, given a dataset like this:

```
geonames_id
1650535
4544379
3469034
```

I want to be able to reconcile the only column there is with Wikidata, and obtain the Qids of the items which bear those particular Geonames identifiers.

## Proposed solution

We redesign the right-hand panel of the reconciliation dialog. Instead of showing the list of all columns except the current one (on which the reconciliation operation is about to run), we display a list of "column mappings" (we could also call it differently). It could look like this:
![image](https://github.com/OpenRefine/OpenRefine/assets/309908/19ab8887-5be2-4ae8-9e1b-c3030c34dc95)

Note that not all columns of the project are displayed: we only list those which are used for reconciliation. We also list the column that is used for the name (in the wireframe above: "Column G"), making it possible to modify or delete this mapping.

When the user clicks "Add", they get a dialog which could look like this:
![image](https://github.com/OpenRefine/OpenRefine/assets/309908/163313bb-4661-4622-9b2d-afb393da60b1)

This lets them configure reconciliation so that a new column is used, either as the name of the entity or as another property.
A similar (or identical) dialog could be used to edit an existing column mapping.

When the reconciliation dialog is opened, initially we could have a single column mapping, for the column the reconciliation operation is run on, matched against the entity name:
![image](https://github.com/OpenRefine/OpenRefine/assets/309908/ce102ee4-cf34-4fc0-9110-1982f6639d21)

This would mean that by default, we are doing the same thing as currently. In the example above where there is only one column in the project, containing Geonames IDs, the user could simply click "Edit" change the column mapping not to be used as an entity name but a property:
![image](https://github.com/OpenRefine/OpenRefine/assets/309908/1d551849-e4e0-4d46-8453-b17376082265)

In the future, this dialog could be extended to contain additional settings for the column mapping: for instance, configuring how the matching is done (should this act as a filter, meaning that only matching entities should be returned, or should it rather boost the score if there is a match, but we should not exclude mis-matching entities?). See https://github.com/reconciliation-api/specs/pull/131 which proposes to add support for such settings in the protocol.

## Alternatives considered

Surely other UIs can also solve this problem.

## Additional context

This new UI would also solve other problems:
* reconciling with multiple names: say I have a column with Swedish names and another one with Finnish names, and both of them should be matched against the name of the entity in the service: we can simply add two column mappings to do that.
* using the same column for matching against different properties (not sure how common this is, though)
* potentially, the "Add" dialog could also be adapted to cater for #2075

**Comment by lozanaross** [comment]
Thanks for this, I think the proposal in principle makes a lot of sense and can improve several confusing issues at the moment in one go. But I just wonder about this:

> Instead of showing the list of all columns except the current one (on which the reconciliation operation is about to run), we display a list of "column mappings" (we could also call it differently).

What would be the starting point here - before users have added any mappings ? If empty, I suspect users might not be really prompted to add anything (because we are not setting any expectations). I also wonder about the whole idea of associating columns with properties (this is something I struggle explaining to new users during tutorials), it can be a tricky concept to get your head around.  

@Lydiaofficial - this is something for you think about, how to develop the concept proposed above into a more user-friendly UI. You can also discuss during your meetings with experienced users this week.

**Comment by wetneb** [comment]
> What would be the starting point here - before users have added any mappings ?

@lozanaross See the third image of my post above, and the text immediately before and after it:
> When the reconciliation dialog is opened, initially we could have a single column mapping, for the column the reconciliation operation is run on, matched against the entity name. This would mean that by default, we are doing the same thing as currently.

**Comment by lozanaross** [comment]
> @lozanaross See the third image of my post above, and the text immediately before and after it:

Haha, this is why I should not comment on issues from my phone while on a conference! Didn't properly scroll through! I think in principle this can work, but I guess the confusing part (from perspective of heavy Wiki use is that the "name" is just the "label" right? and Wikidata / Wikibase users wouldn't think of that as a property per se... (which it is of course in the RDF output, but let's imagine we're working with GLAM Wikidata users who don't look at RDF). 

Maybe instead of "Property mappings" we call the heading: "Column-to-property mappings" or even more generic "Data model mappings" (maybe @Lydiaofficial can query this with users during her chats with them). Then perhaps "name" can be a service-specific term, so for WD / WB "label" would be a better fit. And we could even consider instead of an "Add button" - a dropdown selector below the default mapping. I need to mock this up - will try later!

**Comment by wetneb** [comment]
Yes the "Property mapping" name is not a great one. Maybe just "Column mapping".

Yes, for a Wikibase the "name" is currently a "label", but the reconciliation feature is more general than the Wikibase integration, so I wouldn't want to align all the terminology on Wikibases by principle. Also, even within Wikibase, you could want to reconcile against entities which do not have "labels", for instance lexemes.

**Comment by tfmorris** [comment]
Because this only addresses half of the dialog box, I don't understand how it interacts with the entity type selection on the left side of the box. One of the advantages of having a distinct identifier reconciliation is that it can be separate from the entity type selection which is no longer relevant.

It also doesn't make sense to reconcile against multiple strong identifiers, so I'm wondering how the user is prevented from attempted to doing this. I'm assuming that the reconciliation service will know which properties represent strong identifiers and which are normal values and tailor its behavior appropriately.

**Comment by wetneb** [comment]
> It also doesn't make sense to reconcile against multiple strong identifiers

I actually do this quite often! It is common that I have multiple identifier columns. Those columns might have some blank cells, some cells with values which do not exist on the service (typically, Wikidata is rarely complete), or even some incorrect cells. So I use multiple columns and it's indeed up to the service to decide what to do with the info.

**Comment by wetneb** [comment]
> Because this only addresses half of the dialog box, I don't understand how it interacts with the entity type selection on the left side of the box. One of the advantages of having a distinct identifier reconciliation is that it can be separate from the entity type selection which is no longer relevant.

Yes, I would just leave type selection untouched for now. I think users can decide to use it or not depending on their needs.

**Comment by Abbe98** [comment]
I'm a little worried that this increased complexity would increase the barrier for many users and would question if this type of mappings should be a part of OpenRefine rather than the logic of individual reconciliation services.

Wouldn't it be better if OpenRefine saw reconciliation services as pure search, precision and recall services? A service can map incoming input and fallback to whatever logic that makes sense for that particular service. A service could even consider its input a logical query.

**Wikidata and Wikibase is very much an extreme case and a tricky one but even so wouldn't it be easier to just expose its CirrusSearch capabilities in the reconciliation service rather than to introduce column mappings into the UI?

**Comment by wetneb** [comment]
Do you mean that OpenRefine should only support reconciling one column, without any other additional column? If the user would like to rely on other columns, they would need to concatenate those columns into a single column, formulating an ElasticSearch-like query?

Intuitively that's a pretty big decrease in usability: people would need to learn the expected query syntax of the reconciliation service (which would likely vary from one service to the other) and the GREL expressions to use to formulate the corresponding query in OpenRefine (concatenation, escaping of fields and access of other columns at least).

I don't think my proposal adds a lot of complexity compared to the existing UI: it is just a way to reformat the existing right-hand side area of the dialog. Just like in the current state, user would still be able to ignore this area entirely if they only want to reconciled based on the column they selected earlier. To make it even simpler, we could avoid the extra dialog to configure a new column mapping, and just do that in the area that lists existing column mappings already.

I think your proposal makes sense for other types of reconciliation services, which do not support passing additional properties in a standard way like the reconciliation protocol does. For instance, the OpenSearch services you suggested (https://github.com/OpenRefine/OpenRefine/issues/5472) would fit well with your proposal. I think it would totally be possible to support those as well, offering a different UI when an OpenSearch service is selected. But for the reconciliation protocol it would feel quite at odds with what the protocol offers.

**Comment by nikkiwd** [comment]
I would really _really_ like to have a way to do this. Almost every use I have for OpenRefine involves taking data containing external IDs and loading the Wikidata items which have those IDs, so I can compare the existing data and find where data is missing. Not being able to do that properly is the main reason I don't use OpenRefine very often - I generally don't want to do name-based matching until later, if at all.

I always expect to find this functionality under the column heading > "Reconcile" > "Use values as identifiers". I'd expect to be able to select the property there, and then use "Reconcile" > "Add entity identifiers column…" under the column heading to add a column for the corresponding QID.
